### PR TITLE
nixos/network-interfaces: remove duplicate use of the tempAddresses parameter

### PR DESCRIPTION
**Pull request by Izorkin**
Remove duplicate use of the tempAddresses parameter.
Previously, the `tempAddresses` parameter was applied in two places:
 - `/etc/sysctl.d/60-nixos.conf`
```
...
net.ipv6.conf.default.use_tempaddr=2
net.ipv6.conf.enp1s0.use_tempaddr=2
...
```

 - `/etc/udev/rules.d/98-ipv6-privacy-extensions.rules`
```
# enable and prefer IPv6 privacy addresses by default
ACTION=="add", SUBSYSTEM=="net", RUN+="/nix/store/izpf49b74i15pcr9708s3xdwyqs4jxwl-bash-5.2p32/bin/sh -c 'echo 2 > /proc/sys/net/ipv6/conf/$name/use_tempaddr'"
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
I really have no idea if sysctl or udev is preferred

**Comment by ElvishJerricco** [comment]
Generally I think `boot.kernel.sysctl` is preferred, given it's much more declarative. @Izorkin is there a reason you prefer the udev rule?

Also, are were sure there's not a *reason* it was done in both places? The git blame for these things is a little hard to follow. Is there a problem caused by the way it is now?

**Comment by Izorkin** [comment]
When I boot the server I get this error in the logs:
```
  systemd-sysctl[3450]: Couldn't write '2' to 'net/ipv6/conf/enp3s0/use_tempaddr', ignoring: No such file or directory
```

**Review by ElvishJerricco — CHANGES_REQUESTED** [review]
> When I boot the server I get this error in the logs:

>   `systemd-sysctl[3450]: Couldn't write '2' to 'net/ipv6/conf/enp3s0/use_tempaddr', ignoring: No such file or directory`

That isn't an error. And I'm sure the udev rule is getting the same error but just ignoring it silently. It does the same thing, so why wouldn't it be?

I'm also still not entirely sure there isn't *some* reason it's done in both places.

There's too many unanswered questions.

File: nixos/modules/tasks/network-interfaces.nix [hunk 1]
```diff
@@ -1471,8 +1464,11 @@ in
         text = let
           sysctl-value = tempaddrValues.${cfg.tempAddresses}.sysctl;
         in ''
-          # enable and prefer IPv6 privacy addresses by default
-          ACTION=="add", SUBSYSTEM=="net", RUN+="${pkgs.bash}/bin/sh -c 'echo ${sysctl-value} > /proc/sys/net/ipv6/conf/$name/use_tempaddr'"
+          # Enable and prefer IPv6 privacy addresses by default
+          ACTION=="add", SUBSYSTEM=="net", RUN+="${pkgs.procps}/bin/sysctl -w net.ipv6.conf.all.use_tempaddr=${sysctl-value}"
+          ACTION=="add", SUBSYSTEM=="net", RUN+="${pkgs.procps}/bin/sysctl -w net.ipv6.conf.default.use_tempaddr=${sysctl-value}"
+          ACTION=="add", SUBSYSTEM=="net", RUN+="${pkgs.procps}/bin/sysctl -w net.ipv6.conf.$name.use_tempaddr=${sysctl-value}"
+          ACTION=="add", SUBSYSTEM=="net", RUN+="${pkgs.procps}/bin/sysctl -w net.ipv6.conf.lo.use_tempaddr=-1"
```
[the remark this answers is not in the document]
Comment (Izorkin) [reply nixos/modules/tasks/network-interfaces.nix:1471]: The `net.ipv6.conf.$name.use_tempaddr=${sysctl-value}` rule overwrites the `-1` in interface 'lo' value to `2`. The last rule returns the default value.

**Comment by Izorkin** [comment]
> That isn't an error. And I'm sure the udev rule is getting the same error but just ignoring it silently. It does the same thing, so why wouldn't it be?

I can rewrite to this variant:
```diff
diff --git a/nixos/modules/tasks/network-interfaces.nix b/nixos/modules/tasks/network-interfaces.nix
index c9ea31a85bee..bea5d243612b 100644
--- a/nixos/modules/tasks/network-interfaces.nix
+++ b/nixos/modules/tasks/network-interfaces.nix
@@ -1393,15 +1393,16 @@ in
       "net.ipv6.conf.default.disable_ipv6" = mkDefault (!cfg.enableIPv6);
       # allow all users to do ICMP echo requests (ping)
       "net.ipv4.ping_group_range" = mkDefault "0 2147483647";
-      # networkmanager falls back to "/proc/sys/net/ipv6/conf/default/use_tempaddr"
+      # Enable and prefer IPv6 privacy addresses by default
+      "net.ipv6.conf.all.use_tempaddr" = tempaddrValues.${cfg.tempAddresses}.sysctl;
       "net.ipv6.conf.default.use_tempaddr" = tempaddrValues.${cfg.tempAddresses}.sysctl;
     } // listToAttrs (forEach interfaces
-        (i: nameValuePair "net.ipv4.conf.${replaceStrings ["."] ["/"] i.name}.proxy_arp" i.proxyARP))
-      // listToAttrs (forEach interfaces
         (i: let
           opt = i.tempAddress;
           val = tempaddrValues.${opt}.sysctl;
-         in nameValuePair "net.ipv6.conf.${replaceStrings ["."] ["/"] i.name}.use_tempaddr" val));
+         in nameValuePair "net.ipv6.conf.${replaceStrings ["."] ["/"] i.name}.use_tempaddr" val))
+      // listToAttrs (forEach interfaces
+        (i: nameValuePair "net.ipv4.conf.${replaceStrings ["."] ["/"] i.name}.proxy_arp" i.proxyARP));

     systemd.services.domainname = lib.mkIf (cfg.domain != null) {
       wantedBy = [ "sysinit.target" ];
@@ -1465,16 +1466,6 @@ in
     virtualisation.vswitch = mkIf (cfg.vswitches != { }) { enable = true; };

     services.udev.packages =  [
-      (pkgs.writeTextFile rec {
-        name = "ipv6-privacy-extensions.rules";
-        destination = "/etc/udev/rules.d/98-${name}";
-        text = let
-          sysctl-value = tempaddrValues.${cfg.tempAddresses}.sysctl;
-        in ''
-          # enable and prefer IPv6 privacy addresses by default
-          ACTION=="add", SUBSYSTEM=="net", RUN+="${pkgs.bash}/bin/sh -c 'echo ${sysctl-value} > /proc/sys/net/ipv6/conf/$name/use_tempaddr'"
-        '';
-      })
       (pkgs.writeTextFile rec {
         name = "ipv6-privacy-extensions.rules";
         destination = "/etc/udev/rules.d/99-${name}";
```

**Comment by rnhmjoj** [comment]
> I'm also still not entirely sure there isn't some reason it's done in both places.

I originally added the option in https://github.com/NixOS/nixpkgs/pull/34492 using sysctl. This worked fine, but misleadingly the default of `networking.interfaces.<ifname>.preferTempAddress` only applied to interfaces that were declared with `networking.interfaces.<ifname> = { ... };`, not just any interface.

The udev rules were added later in https://github.com/NixOS/nixpkgs/pull/62835/commits/26317b02aeae3ae5f086f22822ccd9695df03bb8 to allow configured all interfaces, undeclared ones included. That was needed because the global `net.ipv6.conf.all` sysctls don't apply to existing interfaces, rendering it essentially useless as it's identical to `net.ipv6.default`. It's a [kernel bug](https://bugzilla.kernel.org/show_bug.cgi?id=11655) so old people essentially started to take accept it as normal behavior.

I guess that after about 15 years you could say the kernel is not going to fix that, so keeping only the udev rules makes sense, even if it sucks that we have to use udev for this.

**Comment by rnhmjoj** [comment]
I still don't understand what is the purpose of this PR.

Using sysctl.conf is the proper way to set the policy and the udev rules are necessary to work around a kernel bug. We can't rely exclusively on the former, we could use exclutively the latter, but what is the point?
Is this just a cosmetic change?

**Comment by Izorkin** [comment]
Yes, it can be considered a cosmetic solution. But when to fix the problem in Upstream is unknown.
